src/host_beta/messageRouter.ts

```diff
@@ -0,0 +1,242 @@
+            event: 'performance',
+            name: histogramName,
+        });
+    }
+
+    reportError(
+        type: string,
+        message: string,
+        stack: string,
+        filename: string,
+        sourceUrl: string,
+        lineno: number,
+        colno: number): void {
+        // Package up the error info to send to the extension
+        const data = { message, stack, filename, sourceUrl, lineno, colno };
+
+        // Inform the extension of the DevTools telemetry event
+        this.sendTelemetry({
+            data,
+            event: 'error',
+            name: type,
+        });
+    }
+
+    openInEditor(url: string, line: number, column: number, ignoreTabChanges: boolean): void {
+        // Forward the data to the extension
+        const request: IOpenEditorData = { column, line, url, ignoreTabChanges };
+        encodeMessageForChannel(msg => vscode.postMessage(msg, '*'), 'openInEditor', request);
+    }
+
+    getVscodeSettings(callback: (arg0: Record<string, unknown>) => void): void {
+        const id = this.getHostCallbacksNextId++;
+        this.getHostCallbacks.set(id, callback);
+        encodeMessageForChannel(msg => vscode.postMessage(msg, '*'), 'getVscodeSettings', {id});
+    }
+
+    sendToVscodeOutput(consoleMessage: string): void {
+        encodeMessageForChannel(msg => vscode.postMessage(msg, '*'), 'consoleOutput', {consoleMessage});
+    }
+
+    openInNewTab(url: string): void {
+        encodeMessageForChannel(msg => vscode.postMessage(msg, '*'), 'openUrl', {url});
+    }
+
+    sendMessageToBackend(message: string): void {
+        encodeMessageForChannel(msg => vscode.postMessage(msg, '*'), 'websocket', { message });
+    }
+
+    onMessageFromFrame(e: FrameToolsEvent, args: any[]): boolean {
+        switch(e) {
+            case 'openInEditor':
+                const [url, line, column, ignoreTabChanges] = args;
+                this.openInEditor(url, line, column, ignoreTabChanges);
+                return true;
+            case 'openInNewTab':
+                this.openInNewTab(args[0]);
+                return true;
+            case 'recordEnumeratedHistogram':
+                const [actionName, actionCode, bucketSize] = args;
+                this.recordEnumeratedHistogram(actionName, actionCode, bucketSize);
+                return true;
+            case 'recordPerformanceHistogram':
+                const [histogramName, duration] = args;
+                this.recordPerformanceHistogram(histogramName, duration);
+                return true;
+            case 'reportError':
+                const [type, message, stack, filename, sourceUrl, lineno, colno] = args;
+                this.reportError(type, message, stack, filename, sourceUrl, lineno, colno);
+                return true;
+            case 'sendMessageToBackend':
+                const [cdpMessage] = args;
+                this.sendMessageToBackend(cdpMessage);
+                return true;
```

Comment:
Nit: given we're not really validating types here anyway, I'd prefer to reduce the duplication of method names / case statements here. Maybe have the switch cases invoke `encodeMessageForChannel` directly?

Reply:
Latest version of the file has incorporated this change, but I'm not 100% sure I like the look better. can folks take a look and lmk what you think?

Reply:
Yeah, it reads a bit dense. We could flip this around to keep the methods, drop the switch, and use square-bracket notation to pick the method to dispatch to using `.apply` to pass the args. If we go that route I'd prefer to have an object/instance with *only* those methods so we're not exposing access to anything else.

```ts
const messageHandlers = {
  openInEditor(url: string, line: number, column: number, ignoreTabChanges: boolean): void {
    // Forward the data to the extension
    const request: IOpenEditorData = { column, line, url, ignoreTabChanges };
    encodeMessageForChannel(msg => vscode.postMessage(msg, '*'), 'openInEditor', request);
  }
  // ...
}

onMessageFromFrame(e: FrameToolsEvent, args: any[]): boolean {
  const handler = messageHandlers[e];
  if (handler) {
    handler.apply(messageHandlers, args);
    return true;
  }
  return false;
}
```

Reply:
But this was a nit to begin with so feel free to go back to what you had before if you'd prefer ;)

Reply:
Reverted back to this version, as I'm inclined to leave as is. IMO it makes it easier to understand the differences and similarities between the legacy system and this one, as both end up with these functions for encoding messages. I'm open to revisiting this though if folks have strong objections